🤖 Add authors and contributors to Practice Exercises

ErikSchierboom opened the pull request:
With v3, we've introduced the idea of exercise authors and contributors. This information is stored in the exercise's `.meta/config.json` file (see [the docs](https://github.com/exercism/docs/blob/main/anatomy/tracks/practice-exercises.md#file-metaconfigjson)).

Concept Exercises, which are all new, have all defined their authors and contributors. Practice Exercises though are missing this information. In this PR, we attempt to populate the authors and contributors of Practice Exercises based on their commit history.

## Algorithm

We start out by looking at the current Practice Exercises. For each Practice Exercise, we'll look at the commit history of its files to see which commits did something with that Practice Exercise. How renames are handled is discussed in the "Renames" section later in this document. 

Each commit that we can link to a Practice Exercise will be used to determine authorship/contributorship. Here is how we assign authorship/contributorship:

### Authors

1. Find the Git commit author of the oldest commit linked to that Practice Exercise. As there can only be one oldest commit, we only have one author
2. Find the GitHub username for the Git commit author of that commit
3. Add the GitHub username of the Git commit author to the `authors` key in the `.meta/config.json` file

### Contributors

1. Find the Git commit authors of all but the oldest commit linked to that Practice Exercise. If there is only one commit, there won't be any contributors.
1. Exclude the Git commit author of the oldest commit from the list of Git commit authors (an author is _not_ also a contributor) 
2. Find the GitHub usernames for the Git commit authors of those commits
3. Add the GitHub usernames of the Git commit authors to the `contributor` key in the `.meta/config.json` file

We're using the GitHub GraphQL API to find the username of a commit author. In some cases though, a username cannot be found for a commit (e.g. due to the user account no longer existing), in which case we'll skip the commit. 

## Renames

There is a select number of Practice Exercises that might have been renamed at some point. You can ask Git to "follow" a file over its renames using `git log --follow <file>`, which will also return commits made before renames. Unfortunately, Git does not store renames, it just stores the contents of the renamed files and tries to guess if a file was renamed by looking at its contents. This _can_ (and will) lead to false positives, where Git will think a file has been renamed whereas it hasn't. As we don't want to have incorrect authors/contributors for exercises, we're ignoring renames. The only exception to this are known exercise renames:

- `bracket-push` was renamed to `matching-brackets`
- `retree` was renamed to `satellite`
- `resistor-colors` was renamed to `resistor-color-duo`
- `kindergarden-garden` was renamed to `kindergarten-garden`

If you know of a rename of a Practice Exercise, please check to see if its authors and contributors are correctly set.

## Exclusions

There are some commits that we skip over, which thus don't influence the authors/contributors list:

- Commits authored by `dependabot[bot]`, `dependabot-preview[bot]` or `github-actions[bot]`
- Bulk update PRs made by `ErikSchierboom` or `kytrinx` to update the track

## Maintainer TODO list

These are the things you, the maintainers, should do:

- Check the authors and contributors, adding/removing users where the data is missing or incorrect
- Double-check any renamed Practice Exercises
- Check for Practice Exercises with missing authors, which is most likely due to the author not being on GitHub anymore 
- Add additional authors if you know that a Practice Exercise has actually been created by more than one author 

If you find something needs updating, just push a new commit to this PR.

## Tracking

https://github.com/exercism/v3-launch/issues/24

angelikatyborska commented on exercises/practice/bank-account/.meta/config.json:
```diff
@@ -1,6 +1,65 @@
 {
   "blurb": "Simulate a bank account supporting opening/closing, withdraws, and deposits of money. Watch out for concurrent transactions!",
   "authors": [],
```
@ErikSchierboom Empty? Is that allowed?

Here's the commit that adds the exercise: https://github.com/exercism/elixir/commit/0afa70cc5e6affc4f24bf2eb3158a4ca30942fbe#diff-8e71c6ee772e27e4a11c87370b8d2689457a4287af7fb2947fd875093a657d93

Looks like the author doesn't have a github account? I guess that's why it's empty?

angelikatyborska commented on exercises/practice/custom-set/.meta/config.json:
```diff
@@ -1,6 +1,47 @@
 {
   "blurb": "Create a custom set type.",
   "authors": [],
```
Another empty one, added in https://github.com/exercism/elixir/commit/fbbfd1c4f0c4ce936b866f66a2f68cedf9a030cc

angelikatyborska commented on exercises/practice/dot-dsl/.meta/config.json:
```diff
@@ -1,6 +1,47 @@
 {
   "blurb": "Write a Domain Specific Language similar to the Graphviz dot language",
   "authors": [],
```
Another empty one, added in https://github.com/exercism/elixir/commit/cfb5f137f7d10f474a8870497ecf6678c0487c8f

angelikatyborska commented on exercises/practice/forth/.meta/config.json:
```diff
@@ -1,6 +1,59 @@
 {
   "blurb": "Implement an evaluator for a very simple subset of Forth",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/a3382372a1ed71550466844cd310af18645fc340

angelikatyborska commented on exercises/practice/hello-world/.meta/config.json:
```diff
@@ -1,6 +1,53 @@
 {
   "blurb": "The classical introductory exercise. Just say \"Hello, World!\"",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/0b77b05fd7a97b7819f29c0d32e436e600092607#diff-455eb2dc457644ff50c312dd8d72cb57e92595d26f13f5c9a94bfd9e78b4a36c by Rob Phoenix (other person then the other empty ones, but also without a github account)

angelikatyborska commented on exercises/practice/list-ops/.meta/config.json:
```diff
@@ -1,6 +1,53 @@
 {
   "blurb": "Implement basic list operations",
   "authors": [],
```
Another empty one, added in https://github.com/exercism/elixir/commit/7e23cf93269ca6a4b31176b9fbd79918253387f3#diff-639ed1e0b89310f36410cd8377faf89a9d2c9ecc438fbb15b2657b237aefdcc1

angelikatyborska commented on exercises/practice/markdown/.meta/config.json:
```diff
@@ -1,6 +1,35 @@
 {
   "blurb": "Refactor a Markdown parser",
   "authors": [],
```
Another empty one, added in https://github.com/exercism/elixir/commit/b4dac08e9dde0956bac0a994827e2c96a89b7451 by Yeong Sheng, Tan

angelikatyborska commented on exercises/practice/minesweeper/.meta/config.json:
```diff
@@ -1,6 +1,59 @@
 {
   "blurb": "Add the numbers to a minesweeper board",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/dcfe347db2c3685954439a72694958dcfe94a3b0

angelikatyborska commented on exercises/practice/ocr-numbers/.meta/config.json:
```diff
@@ -1,6 +1,29 @@
 {
   "blurb": "Given a 3 x 4 grid of pipes, underscores, and spaces, determine which number is represented, or whether it is garbled.",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/93023194ccf7450438df8652c051a5622027a13e by Brian Carrigan

angelikatyborska commented on exercises/practice/parallel-letter-frequency/.meta/config.json:
```diff
@@ -1,6 +1,59 @@
 {
   "blurb": "Count the frequency of letters in texts using parallel computation.",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/1b64e528d2c18f9ec392b239f1107bfa4b53434c

angelikatyborska commented on exercises/practice/poker/.meta/config.json:
```diff
@@ -1,6 +1,29 @@
 {
   "blurb": "Pick the best hand(s) from a list of poker hands.",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/0583fc6d09a4999ca032692018a977f06c052ee9 by Chris Doggett

angelikatyborska commented on exercises/practice/robot-simulator/.meta/config.json:
```diff
@@ -1,6 +1,35 @@
 {
   "blurb": "Write a robot simulator.",
   "authors": [],
```
Another empty one, added in https://github.com/exercism/elixir/commit/ad9029641e5b09315979d125b01ce3c816229877 by Chris Doggett

angelikatyborska commented on exercises/practice/sublist/.meta/config.json:
```diff
@@ -1,6 +1,56 @@
 {
   "blurb": "Write a function to determine if a list is a sublist of another list.",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/6f82339d811367ac110e739d041e8c3ae7ae7b2d

angelikatyborska commented on exercises/practice/zipper/.meta/config.json:
```diff
@@ -1,6 +1,62 @@
 {
   "blurb": "Creating a zipper for a binary tree.",
   "authors": [],
```
another empty one, added in https://github.com/exercism/elixir/commit/d0d4cd185bf7e90328af2299f49acda190fd85b4

angelikatyborska reviewed (approved):
I checked manually the first 20 or so exercises for their authors and your script got all of those correct, so I feel comfortable assuming the rest are also correct.

There is over 10 exercises that have an empty array for authors still, I left a comment for each one.

I only skimmed the contributors, but they make sense to me. The track maintainers show up in almost every exercise. I checked some names that stand out and they did make a few commits for those exercises where they show up.

If the empty array for authors is not a problem, then I think it's good to merge.

ErikSchierboom commented:
> There is over 10 exercises that have an empty array for authors still, I left a comment for each one.

> @ErikSchierboom Empty? Is that allowed?

No, that isn't allowed currently. We might want to reconsider that for practice exercises though.

> Looks like the author doesn't have a github account? I guess that's why it's empty?

That is correct. As an example, let's double-check this commit: https://github.com/darktef/elixir-exercism/commit/d0d4cd185bf7e90328af2299f49acda190fd85b4 If we request the author information for that commit via the following graphql query

```graphql
query { 
 node(id: "MDY6Q29tbWl0MTIwMTYzNzgwOmQwZDRjZDE4NWJmN2U5MDMyOGFmMjI5OWY0OWFjZGExOTBmZDg1YjQ=") {
  ... on Commit {
    author {
      name
      user {
        login
      }
    }
  }
}
}
```

we get:

```json
{
  "data": {
    "node": {
      "author": {
        "name": "Peter Minten",
        "user": null
      }
    }
  }
}
```

If we do the same for [this commit](https://github.com/exercism/elixir/commit/9f7882ff18155c3ed36466b95f1a1be6e51baefd), made by you, we get:

```json
{
  "data": {
    "node": {
      "author": {
        "name": "Angelika Tyborska",
        "user": {
          "login": "angelikatyborska"
        }
      }
    }
  }
}
```

So yes: in some cases a user does not have a github account, which is why the authors array can be empty.

> I only skimmed the contributors, but they make sense to me. The track maintainers show up in almost every exercise. I checked some names that stand out and they did make a few commits for those exercises where they show up.

Yeah, there are usually some track-wide PRs that cause people to be a contributor to all the exercises. If a track doesn't want these to count towards being a contributor (which sort of makes sense), this has to be done manually.

ErikSchierboom committed:
Add authors and contributors to Practice Exercises

For each Practice Exercise, we've looked at the commit history of its files to see which commits touched that Practice Exercise.

Each commit that we could associate with a Practice Exercise was used to determine authorship/contributorship.

### Authors

1. Find the Git commit author of the oldest commit linked to that Practice Exercise
2. Find the GitHub username for the Git commit author of that commit
3. Add the GitHub username of the Git commit author to the `authors` key in the `.meta/config.json` file

### Contributors

1. Find the Git commit authors and any co-authors of all but the oldest commit linked to that Practice Exercise. If there is only one commit, there won't be any contributors.
1. Exclude the Git commit author and any co-authors of the oldest commit from the list of Git commit authors (an author is _not_ also a contributor)
2. Find the GitHub usernames for the Git commit authors and any co-authors of those commits
3. Add the GitHub usernames of the Git commit authors and any co-authors to the `contributor` key in the `.meta/config.json` file

We used the GitHub GraphQL API to find the username of a commit author or any co-authors. In some cases though, a username cannot be found for a commit (e.g. due to the user account no longer existing), in which case the commit was skipped.

## Renames

There are a small number of Practice Exercises that might have been renamed at some point. You can ask Git to "follow" a file over its renames using `git log --follow <file>`, which will also return commits made before renames. Unfortunately, Git does not store renames, it just stores the contents of the renamed files and tries to guess if a file was renamed by looking at its contents. This _can_ (and will) lead to false positives, where Git will think a file has been renamed whereas it hasn't. As we don't want to have incorrect authors/contributors for exercises, we're ignoring renames. The only exception to this are known exercise renames:

- `bracket-push` was renamed to `matching-brackets`
- `retree` was renamed to `satellite`
- `resistor-colors` was renamed to `resistor-color-duo`
- `kindergarden-garden` was renamed to `kindergarten-garden`

## Exclusions

There are some commits that we skipped over, and which thus didn't influence the authors/contributors list:

- Commits authored by `dependabot[bot]`, `dependabot-preview[bot]` or `github-actions[bot]`
- Bulk update PRs made by `ErikSchierboom` or `kytrinx` to update the track

angelikatyborska committed:
Remove Billiam from gigasecond contributors
